### PR TITLE
network: stateless vote compression

### PR DESCRIPTION
**Pull request by cce**
## Summary

Similar to #4589  this provides single-message vote compression without any shared context between messages. It removes msgpack control characters and field names, reducing the size of votes down to just the values present, plus a 2-byte header. It provides 25% smaller messages on typical votes, compared to 6-7% smaller for zstd, and is >20x faster than zstd's fastest levels. 

## Test Plan

- Adds options to protocol.RandomizeObject to use for testing encoding/decoding random and 10% empty values (used from vpack_test.go).
- Also uses rapid to generate random votes, wiring them up to regular tests and Go fuzz tests.
- Adds test hook to agreement to check all invocations of makeVote when run during tests
- ~~Adds features to algodump for benchmarking and testing against network data~~ (will submit as separate PR)
- Some testing based on plan from #4589:
  - [x] Unit test checking old + new combinations
  - [x] Manual TwoNodes50Each test with master (m) + feature (f)
  - [x] Manual ThreeNodesEvenDist test with [m]m+f, m+f[f] where [r] is a relay node
  - [x] Manual ThreeNodesEvenDist test with relay running feature and other two nodes on master
  - [x] Perf test

### CODE REVIEW

File: agreement/vote_test.go [hunk 1]
```diff
@@ -27,10 +31,40 @@ import (
...
 	"github.com/algorand/go-algorand/logging"
+	"github.com/algorand/go-algorand/network/vpack"
 	"github.com/algorand/go-algorand/protocol"
 	"github.com/algorand/go-algorand/test/partitiontest"
 )
 
+func init() {
+	testMakeVoteCheckMu.Lock()
+	testMakeVoteCheck = testVPackMakeVote
+	testMakeVoteCheckMu.Unlock()
```
Comment (jannotti) [inline agreement/vote_test.go:42]: What could be contending at init time?

File: network/vpack/gen.go [hunk 2]
```diff
@@ -0,0 +1,660 @@
...
+	g.parseFns[t] = true
+
+	fields := exportedCodecFields(t)
+	sort.Slice(fields, func(i, j int) bool {
+		return getCodecTagName(fields[i]) < getCodecTagName(fields[j])
+	})
+
+	// Check for fixed-size structs using the vpack_size tag on _struct field
+	fixedSize := 0
+	// Look for _struct field directly (which may be unexported)
```
Comment (jannotti) [inline network/vpack/gen.go:346]: "may" be?
It is unexported by definition, isn't it?

**Review by jannotti — COMMENTED** [review]
Couple questions so far.

File: tools/debug/algodump/compress.go [hunk 3]
```diff
@@ -0,0 +1,601 @@
...
+// Compression-specific flags
+var compress = flag.Bool("compress", false, "Compress messages and show statistics")
+var dictionaryFile = flag.String("dict", "", "Path to zstd dictionary file to use for compression")
+var compressionLevel = flag.Int("level", 3, "Compression level (1-9, higher = better compression but slower)")
+var printCompressGrow = flag.Bool("printCompressGrow", false, "Print hex dump of messages that grow after compression")
+var storeDir = flag.String("store", "", "Directory to store messages in")
+var storeBatchSize = flag.Int("storeBatch", 1024*1024, "Flush messages to disk after accumulating this many bytes")
+var compressionWindowSize = flag.Int("windowSize", 32768, "Compression window size in bytes")
+var contextConfig = flag.String("contextcfg", "", "Configure tag-specific compression contexts with window sizes. Format: TAG1:size1;TAG2,TAG3:size2 (e.g. AV:4096;TX,PP:32768)")
+var useGozstd = flag.Bool("useGozstd", false, "Use gozstd (C implementation) instead of klauspost (Go implementation) for compression contexts")
```
Comment (gmalouf) [inline tools/debug/algodump/compress.go:52]: Long term, are we going to want to keep support for both compression libraries?

File: network/vpack/parse_test.go [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+	"github.com/algorand/go-algorand/crypto"
+	"github.com/algorand/go-algorand/protocol"
+	"github.com/algorand/go-algorand/test/partitiontest"
+	"github.com/stretchr/testify/require"
+)
+
+// a string that is greater than the max 5-bit fixmap size
+const gtFixMapString = "12345678901234567890123456789012"
+
+var parseVoteTestCases = []struct {
```
Comment (gmalouf) [inline network/vpack/parse_test.go:32]: Wondering if on top of these static test cases, whether its worth doing some property-based generated cases using pgregory.net/rapid as in `wsnetwork_test.go`.

File: protocol/codec_tester.go [hunk 5]
```diff
@@ -49,27 +49,55 @@ func oneOf(n int) bool {
 	return (rand.Int() % n) == 0
 }
 
+type randomizeObjectCfg struct {
```
Comment (gmalouf) [inline protocol/codec_tester.go:52]: Maybe rapid would make sense for this as well - does similar behavior.

File: network/vpack/vpack.go [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+			pos += 8
+		case markerLiteralBin64:
+			if pos+63 >= lenb { // Needs 64 more bytes
+				return nil, io.ErrUnexpectedEOF
+			}
+			dst = append(dst, msgpBin8Len64...)
+			dst = append(dst, src[pos:pos+64]...)
+			pos += 64
+		case markerLiteralBin80:
+			if pos+79 >= lenb { // Needs 80 more bytes
```
Comment (gmalouf) [inline network/vpack/vpack.go:181]: Is it worth enhancing tests to cover each of these failure cases?

**Review by gmalouf — COMMENTED** [review]
Left a few comments, mostly follow the approach. I'm worried about covering corner cases during vote compression, so was happy to see the fuzzer on vote fields encoding tests.

File: network/vpack/msgp.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+
+package vpack
+
+import (
+	"fmt"
+)
+
+// Minimal msgpack constants used here
+const (
+	fixMapMask = 0x80
```
Comment (algorandskiy) [inline network/vpack/msgp.go:25]: I wonder if we can import them from msgp lib?

File: network/vpack/msgp.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+// Zero-allocation string reading that returns a slice of the original data
+func (p *parser) readString() ([]byte, error) {
+	b, err := p.readByte()
+	if err != nil {
+		return nil, err
+	}
+	if b < fixStrMask || b > fixStrMax {
+		return nil, fmt.Errorf("readString: expected fixstr, got 0x%02x", b)
```
Comment (algorandskiy) [inline network/vpack/msgp.go:87]: > // Zero-allocation string reading that returns a slice of the original data

so format string makes things worse
```
network/vpack/msgp.go:87:69: b escapes to heap
```

```suggestion
	if b < fixStrMask || b > fixStrMax {
		val := b
		return nil, fmt.Errorf("readString: expected fixstr, got 0x%02x", val)
```
In this case `val` will escape only in case of an error making this function indeed zero allocs. Btw, the same applies to `readFixMap`, `readUintBytes` and `ensureBytes` where `n` escapes

**Review by algorandskiy — COMMENTED** [review]
Looks reasonable and surprising that encoding only field names gave 20% of saving

File: network/vpack/msgp.go (hunk 8 above)
Comment (algorandskiy) [reply network/vpack/msgp.go:87]: not really, when compiler determines a var escapes, its allocation goes to heap not stack/regs.

Comment (algorandskiy) [reply network/vpack/msgp.go:87]: I found some clues in x86 assembly
```
                return nil, fmt.Errorf("readString: expected fixstr, got 0x%02x", b)
  0xba4d0a              488d150ffe2500          LEAQ 0x25fe0f(IP), DX
  0xba4d11              4889542430              MOVQ DX, 0x30(SP)
  0xba4d16              0fb6d0                  MOVZX AL, DX
  0xba4d19              4c8d05e014b200          LEAQ runtime.staticuint64s(SB), R8
```
and
```
// staticuint64s is used to avoid allocating in convTx for small integer values.
// staticuint64s[0] == 0, staticuint64s[1] == 1, and so forth.
// It is defined in assembler code so that it is read-only.
var staticuint64s [256]uint64
```

so it appears to be some later optimization, LTO?

**Review by algorandskiy — DISMISSED** [review]

**Review by gmalouf — DISMISSED** [review]

**Review by algorandskiy — DISMISSED** [review]

File: crypto/onetimesig.go [hunk 9]
```diff
@@ -37,24 +37,24 @@ type OneTimeSignature struct {
...
-	Sig ed25519Signature `codec:"s"`
+	Sig ed25519Signature `codec:"s" vpack:"literal"`
 	PK  ed25519PublicKey `codec:"p"`
 
 	// Old-style signature that does not use proper domain separation.
 	// PKSigOld is unused; however, unfortunately we forgot to mark it
 	// `codec:omitempty` and so it appears (with zero value) in certs.
 	// This means we can't delete the field without breaking catchup.
-	PKSigOld ed25519Signature `codec:"ps"`
+	PKSigOld ed25519Signature `codec:"ps" vpack:"literal" vpack_special_values:"0"`
```
Comment (jannotti) [inline crypto/onetimesig.go:50]: I'm ok with it if we don't, but do we manage to eliminate this entirely, because the key and (empty) value are all one string leading right into the next string, `p2`?  Or do we end up with one byte that represents `ps=0`, followed by something for `p2`?

File: network/vpack/parse.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+
+import (
+	"fmt"
+)
+
+func parseVote(data []byte, c compressWriter) error {
+	p := newParser(data)
+
+	// Fixed size struct with 3 fields
+	cnt, err := p.readFixMap()
```
Comment (jannotti) [inline network/vpack/parse.go:13]: The fields are `omitempty`, but here we assume they will never be omitted. I wonder if we should remove the `omitempty`. I think it's true that they can never be empty.  I'm almost like a msgpack `failempty` tag.

**Review by jannotti — APPROVED** [review]
I have some thoughts, but nothing stopping approval.

File: network/vpack/parse.go [hunk 11]
```diff
@@ -0,0 +1,292 @@
+// Code generated by gen.go; DO NOT EDIT.
```
Comment (algorandskiy) [inline network/vpack/parse.go:1]: shouldn't we add `go:generate` into gen.go in order to force build process the gen source to its product verification?

File: network/vpack/parse.go [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+
+		case "r":
+			c.writeStatic(staticIdxRField)
+
+			// Variable size struct
+			cnt, err := p.readFixMap()
+			if err != nil {
+				return fmt.Errorf("reading map for rawVote: %w", err)
+			}
+			if cnt < 1 || cnt > 5 {
```
[the remark this answers is not in the document]
Comment (jannotti) [reply network/vpack/parse.go:72]: Ok, so `cnt` must be at least 2?

File: network/vpack/parse.go [hunk 13]
```diff
@@ -0,0 +1,292 @@
...
+		return fmt.Errorf("reading map for unauthenticatedVote: %w", err)
+	}
+	if cnt != 3 {
+		return fmt.Errorf("expected fixed map size 3 for unauthenticatedVote, got %d", cnt)
+	}
+	c.writeStatic(staticIdxMapMarker3)
+
+	for range 3 {
+
+		key, err := p.readString()
```
[the remark this answers is not in the document]
Comment (jannotti) [reply network/vpack/parse.go:24]: What is the usual size of a compressed vote with the current changes?  I doubt 8-9 bytes is worth it. I'm guessing votes are in the 200 byte range?

File: crypto/onetimesig.go (hunk 9 above)
Comment (jannotti) [reply crypto/onetimesig.go:50]: I think we probably take the big win we've got and not worry about this or the other place with the 8-9 byte idea.  But nice to have the conversation here for posterity.  I suppose it matters more as stake is more decentralized - more votes per round.

**Review by jannotti — DISMISSED** [review]

File: network/vpack/parse.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (jannotti) [reply network/vpack/parse.go:24]: I meant after vpack compression as it stands now.

**Review by algorandskiy — DISMISSED** [review]

**Review by gmalouf — DISMISSED** [review]

File: network/vpack/parse.go [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+	rStepVoteValue
+	sigP1sVoteValue
+	sigP2VoteValue
+	sigP2sVoteValue
+	sigPVoteValue
+	sigPsVoteValue
+	sigSVoteValue
+)
+
+func parseVote(data []byte, c *StatelessEncoder) error {
```
Comment (algorandskiy) [inline network/vpack/parse.go:41]: need a comment stating `parseVote` parses msgp and produces encoding vote by using `StatelessEncoder`

**Review by algorandskiy — APPROVED** [review]
LGMT. Two points:
1. P2P implementation misses peer feature headers
2. 2/3 of the PR is messagepack business (reading and writing). How about changing the spec and let agreement emitting compressed version out of go struct directly? Technically it does not matter much msgp or vpack encoding is used

File: network/vpack/msgp.go [hunk 15]
```diff
@@ -0,0 +1,209 @@
...
+		return nil, err
+	}
+	s := p.data[p.pos : p.pos+length]
+	p.pos += length
+	return s, nil
+}
+
+func (p *msgpVoteParser) readBin80() ([80]byte, error) {
+	var data [80]byte
+	if err := p.ensureBytes(2); err != nil {
```
Comment (jannotti) [inline network/vpack/msgp.go:125]: You `ensureBytes(82)` here and get it done early?

File: network/vpack/parse.go (hunk 14 above)
Comment (jannotti) [reply network/vpack/parse.go:41]: Maybe rename `data` to `msgpacked`?

File: network/vpack/vpack.go [hunk 16]
```diff
@@ -0,0 +1,384 @@
...
+// presence is indicated in a 1-byte bitmask in the header.
+const (
+	bitPer    uint8 = 1 << iota // r.per
+	bitDig                      // r.prop.dig
+	bitEncDig                   // r.prop.encdig
+	bitOper                     // r.prop.oper
+	bitOprop                    // r.prop.oprop
+	bitStep                     // r.step
+
+	propFieldsMask      uint8 = bitDig | bitEncDig | bitOper | bitOprop
```
Comment (jannotti) [inline network/vpack/vpack.go:37]: why doesn't `bitPer` get or'd in?

File: network/vpack/vpack.go [hunk 17]
```diff
@@ -0,0 +1,384 @@
...
+}
+
+// NewStatelessEncoder returns a new StatelessEncoder.
+func NewStatelessEncoder() *StatelessEncoder {
+	return &StatelessEncoder{}
+}
+
+// CompressVoteBound estimates the maximum size needed for compressed vote data
+func CompressVoteBound(srcSize int) int {
+	// Vote messages typically get ~80% smaller, but we will use
```
Comment (jannotti) [inline network/vpack/vpack.go:61]: I guess it can't possibly matter, but you _know_ it will get smaller by at least removing all the msgpack names, right?

**Review by jannotti — APPROVED** [review]
One question I care about, the rest are small thoughts.

**Review by gmalouf — APPROVED** [review]

**Review by gmalouf — DISMISSED** [review]

**Review by jannotti — DISMISSED** [review]

**Review by algorandskiy — DISMISSED** [review]

**Review by gmalouf — DISMISSED** [review]

**Review by gmalouf — DISMISSED** [review]

**Review by jannotti — DISMISSED** [review]

**Review by jannotti — APPROVED** [review]

**Review by gmalouf — DISMISSED** [review]

File: network/wsNetwork.go [hunk 18]
```diff
@@ -302,6 +302,8 @@ type msgBroadcaster struct {
 	broadcastQueueBulk     chan broadcastRequest
 	// slowWritingPeerMonitorInterval defines the interval between two consecutive tests for slow peer writing
 	slowWritingPeerMonitorInterval time.Duration
+	// enableVoteCompression controls whether vote compression is enabled
+	enableVoteCompression bool
```
Comment (algorandskiy) [inline network/wsNetwork.go:306]: usage of this `enableVoteCompression` vs `wn.config. EnableVoteCompression` is not consistent, maybe `wn.enableVoteCompression` should be used everywhere

**Review by algorandskiy — DISMISSED** [review]

**Review by algorandskiy — APPROVED** [review]

**Review by jannotti — APPROVED** [review]

**Review by gmalouf — APPROVED** [review]
